File: 18 - Extending RAD Server with Resource Interceptors/C++/advanced/BaseInterceptor.cpp
//---------------------------------------------------------------------------

#pragma hdrstop

#include "BaseInterceptor.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

// Helper methods
System::String TBaseInterceptor::GenerateRequestID()
{
	return FormatDateTime("yyyymmdd-hhnnss-", Now()) + IntToStr(Random(9999));
}

System::String TBaseInterceptor::GetHeaderValue(TEndpointRequest* ARequest, const System::String& AHeaderName)
{
	try {
		return ARequest->Headers->GetValue(AHeaderName);
	}
	catch (Exception& E) {
		// Header not found or other error
		return "";
	}
}

System::String TBaseInterceptor::ExtractAuthToken(TEndpointRequest* ARequest)
{
	// Use custom header to avoid conflicts with RAD Server's built-in auth
	System::String AuthHeader = GetHeaderValue(ARequest, "X-Custom-Token");
	if (AuthHeader != "") {
		return AuthHeader;
	}
	else {
		return "";
	}
}

bool TBaseInterceptor::ValidateAuthToken(const System::String& AToken)
{
	// Simple validation - in real app, validate JWT or check database
	return AToken != "";
	// For demo purposes, accept any non-empty token
}

System::String TBaseInterceptor::GetUserIDFromToken(const System::String& AToken)
{
	// Simple extraction - in real app, decode JWT and extract user ID
	if (AToken.Pos("user:") > 0) {
		return AToken.SubString(AToken.Pos("user:") + 5, AToken.Length());
	}
	else if (AToken.Pos("admin") > 0) {
		return "admin-user";
	}
	else if (AToken != "") {
		return "user-" + AToken.SubString(1, Min(10, AToken.Length())); // Use first 10 chars of token
	}
	else {
		return "anonymous";
	}
}

// Check if metadata should be included based on URL parameter
bool TBaseInterceptor::ShouldIncludeMetadata(TEndpointRequest* ARequest)
{
	try {
		System::String MetadataParam = ARequest->Params->Values["metadata"];
		if (MetadataParam == "false" || MetadataParam == "0" || MetadataParam == "no") {
			return false;
		}
		// Default to true for any other value or if parameter is missing
		return true;
	}
	catch (Exception& E) {
		// If there's any error reading the parameter, default to including metadata
		return true;
	}
}

void TBaseInterceptor::LogRequestStart(TEndpointRequest::TMethod AMethod, const System::String& APath, const System::String& AUserID)
{
	// Convert method enum to string for logging
	System::String MethodStr;
	switch (AMethod) {
		case TEndpointRequest::TMethod::Get: MethodStr = "GET"; break;
		case TEndpointRequest::TMethod::Put: MethodStr = "PUT"; break;
		case TEndpointRequest::TMethod::Post: MethodStr = "POST"; break;
		case TEndpointRequest::TMethod::Head: MethodStr = "HEAD"; break;
		case TEndpointRequest::TMethod::Delete: MethodStr = "DELETE"; break;
		case TEndpointRequest::TMethod::Patch: MethodStr = "PATCH"; break;
		case TEndpointRequest::TMethod::Other: MethodStr = "OTHER"; break;
		default: MethodStr = "UNKNOWN"; break;
	}
	
	TLogHelpers::LogMessage(Format("[%s] %s %s - User: %s - Started",
		ARRAYOFCONST((FRequestID, MethodStr, APath, AUserID))));
}

void TBaseInterceptor::LogRequestComplete(int AStatusCode, int ADuration)
{
	TLogHelpers::LogMessage(Format("[%s] Completed in %dms with status %d",
		ARRAYOFCONST((FRequestID, ADuration, AStatusCode))));
}

// IEMSResourceInterceptor implementation
void __fastcall TBaseInterceptor::BeforeRequest(TEndpointContext* AContext, TEndpointRequest* ARequest, TEndpointResponse* AResponse, bool& AHandled)
{

	// Initialize request tracking
	FRequestID = GenerateRequestID();
	FStartTime = Now();

	// Authentication
	System::String Token = ExtractAuthToken(ARequest);
	bool IsAuthenticated;
	System::String UserID;
	DoCustomAuthentication(ARequest, IsAuthenticated, UserID);

	if (!IsAuthenticated) {
		TJSONObject* ErrorResponse = new TJSONObject();
		if (Token == "") {
			ErrorResponse->AddPair("error", "Authentication required");
			ErrorResponse->AddPair("message", "Missing X-Custom-Token header");
		}
		else {
			ErrorResponse->AddPair("error", "Authentication failed");
			ErrorResponse->AddPair("message", "Invalid custom token");
		}

		AResponse->StatusCode = 401;
		AResponse->Body->SetValue(ErrorResponse, true);
		AHandled = true;
		return;
	}

	FCurrentUserID = UserID;

	// Custom validation
	bool IsValid;
	System::String ErrorMessage;
	IsValid = true;
	DoCustomValidation(ARequest, IsValid, ErrorMessage);
	if (!IsValid) {
		TJSONObject* ErrorResponse = new TJSONObject();
		ErrorResponse->AddPair("error", "Validation failed");
		ErrorResponse->AddPair("message", ErrorMessage);

		AResponse->StatusCode = 400;
		AResponse->Body->SetValue(ErrorResponse, true);
		AHandled = true;
		return;
	}

	// Log request start - use ARequest->MethodString directly
	LogRequestStart(ARequest->Method, ARequest->BasePath, FCurrentUserID);
}

void __fastcall TBaseInterceptor::AfterRequest(TEndpointContext* AContext, TEndpointRequest* ARequest, TEndpointResponse* AResponse)
{
	int Duration = MilliSecondsBetween(Now(), FStartTime);

	// Check if metadata should be included
	if (!ShouldIncludeMetadata(ARequest)) {
		// Just log completion without metadata
		LogRequestComplete(200, Duration); // Default assumption
		DoCustomLogging(ARequest, AResponse);
		return;
	}

	// Add metadata to response
	try {
		// Create new response structure
		TJSONObject* ResponseObj = new TJSONObject();

		// Get the original response data directly from AResponse.Body
		TJSONValue* ResponseValue = nullptr;
		if (AResponse->Body->TryGetValue(ResponseValue)) {
			ResponseObj->AddPair("data", ResponseValue);
		}
		else {
			ResponseObj->AddPair("data", new TJSONNull());
		}

		// Add metadata
		TJSONObject* MetaObj = new TJSONObject();
		MetaObj->AddPair("request_id", FRequestID);
		MetaObj->AddPair("timestamp", DateTimeToStr(Now()));
		MetaObj->AddPair("duration_ms", new TJSONNumber(Duration));
		MetaObj->AddPair("user_id", FCurrentUserID);
		MetaObj->AddPair("method", ARequest->MethodString);
		MetaObj->AddPair("path", ARequest->BasePath);

		ResponseObj->AddPair("meta", MetaObj);

		// Set the new response body - maintain ownership to prevent RAD Server conflicts
		AResponse->Body->SetValue(ResponseObj, false);
	}
	catch (Exception& E) {
		// If something goes wrong, just log it but don't break the response
		TLogHelpers::LogMessage(Format("[%s] Error adding metadata: %s", ARRAYOFCONST((FRequestID, E.Message))));
	}

	// Log request completion
	LogRequestComplete(200, Duration); // Default assumption

	// Allow descendants to add custom logging
	DoCustomLogging(ARequest, AResponse);
}

// Default implementations - descendants can override these
void TBaseInterceptor::DoCustomAuthentication(TEndpointRequest* ARequest, bool& AAuthenticated, System::String& AUserID)
{
	System::String Token = ExtractAuthToken(ARequest);
	AAuthenticated = ValidateAuthToken(Token);
	if (AAuthenticated) {
		AUserID = GetUserIDFromToken(Token);
	}
	else {
		AUserID = "";
	}
}

void TBaseInterceptor::DoCustomValidation(TEndpointRequest* ARequest, bool& AValid, System::String& AErrorMessage)
{
	// Default validation - just check content type for POST/PUT
	AValid = true;
	if ((ARequest->Method == TEndpointRequest::TMethod::Post) || (ARequest->Method == TEndpointRequest::TMethod::Put)) {
		System::String ContentType = GetHeaderValue(ARequest, "Content-Type");
		if (ContentType == "") {
			AValid = false;
			AErrorMessage = "Content-Type header is required";
		}
		else if (ContentType != "application/json") {
			AValid = false;
			AErrorMessage = "Content-Type must be application/json";
		}
	}
}

void TBaseInterceptor::DoCustomLogging(TEndpointRequest* ARequest, TEndpointResponse* AResponse)
{
	// Default implementation does nothing - descendants can override
}

File: 18 - Extending RAD Server with Resource Interceptors/Delphi/advanced/BaseInterceptor.pas
unit BaseInterceptor;

interface

uses
  System.SysUtils,
  System.Classes,
  System.JSON,
  System.DateUtils,
  System.TypInfo,
  System.Math,
  EMS.Services,
  EMS.ResourceAPI,
  EMS.ResourceTypes,
  EMSHosting.Helpers;

type
  // Simple base interceptor for RAD Server - demonstrates extending a base class
  TBaseInterceptor = class(TInterfacedPersistent, IEMSResourceInterceptor)
  private
    FRequestID: string;
    FStartTime: TDateTime;

  protected
    FCurrentUserID: string;

    // Helper methods
    function GenerateRequestID: string;
    function ExtractAuthToken(const ARequest: TEndpointRequest): string;
    function ValidateAuthToken(const AToken: string): Boolean;
    function GetUserIDFromToken(const AToken: string): string;
    function GetHeaderValue(const ARequest: TEndpointRequest; const AHeaderName: string): string;
    function ShouldIncludeMetadata(const ARequest: TEndpointRequest): Boolean;
    procedure LogRequestStart(const ARequest: TEndpointRequest; const AUserID: string);
    procedure LogRequestComplete(const AStatusCode: Integer; const ADuration: Integer);

    // Virtual methods that can be overridden by descendants
    procedure DoCustomValidation(const ARequest: TEndpointRequest; var AValid: Boolean; var AErrorMessage: string); virtual;
    procedure DoCustomLogging(const ARequest: TEndpointRequest; const AResponse: TEndpointResponse); virtual;
    procedure DoCustomAuthentication(const ARequest: TEndpointRequest; var AAuthenticated: Boolean; var AUserID: string); virtual;

  public
    constructor Create;
    destructor Destroy; override;

    // Access properties for descendant classes
    property CurrentUserID: string read FCurrentUserID;

    // IEMSResourceInterceptor implementation
    procedure BeforeRequest(const AContext: TEndpointContext; const ARequest: TEndpointRequest; const AResponse: TEndpointResponse; var AHandled: Boolean);
    procedure AfterRequest(const AContext: TEndpointContext; const ARequest: TEndpointRequest; const AResponse: TEndpointResponse);
  end;

implementation

// Instance constructor
constructor TBaseInterceptor.Create;
begin
  // No initialization needed for removed fields
end;


destructor TBaseInterceptor.Destroy;
begin
  inherited;
end;

// Helper methods
function TBaseInterceptor.GenerateRequestID: string;
begin
  Result := FormatDateTime('yyyymmdd-hhnnss-', Now) + IntToStr(Random(9999));
end;


function TBaseInterceptor.GetHeaderValue(const ARequest: TEndpointRequest; const AHeaderName: string): string;
begin
  try
    Result := ARequest.Headers.GetValue(AHeaderName);
  except
    on E: Exception do
    begin
      // Header not found or other error
      Result := '';
    end;
  end;
end;


function TBaseInterceptor.ExtractAuthToken(const ARequest: TEndpointRequest): string;
var
  AuthHeader: string;
begin
  // Use custom header to avoid conflicts with RAD Server's built-in auth
  AuthHeader := GetHeaderValue(ARequest, 'X-Custom-Token');
  if AuthHeader <> '' then
    Result := AuthHeader
  else
    Result := '';
end;


function TBaseInterceptor.ValidateAuthToken(const AToken: string): Boolean;
begin
  // Simple validation - in real app, validate JWT or check database
  Result := AToken <> '';
  // For demo purposes, accept any non-empty token
end;


function TBaseInterceptor.GetUserIDFromToken(const AToken: string): string;
begin
  // Simple extraction - in real app, decode JWT and extract user ID
  if AToken.Contains('user:') then
    Result := AToken.Substring(AToken.IndexOf('user:') + 5)
  else if AToken.Contains('admin') then
    Result := 'admin-user'
  else if AToken <> '' then
    Result := 'user-' + AToken.Substring(0, Min(10, Length(AToken))) // Use first 10 chars of token
  else
    Result := 'anonymous';
end;


function TBaseInterceptor.ShouldIncludeMetadata(const ARequest: TEndpointRequest): Boolean;
var
  MetadataParam: string;
begin
  // Default to true (include metadata) for backward compatibility
  Result := True;
  
  try
    MetadataParam := ARequest.Params.Values['metadata'];
    if MetadataParam <> '' then
    begin
      // Parse metadata parameter - accept 'true', '1', 'yes' as true
      MetadataParam := LowerCase(Trim(MetadataParam));
      Result := (MetadataParam = 'true') or (MetadataParam = '1') or (MetadataParam = 'yes');
    end;
  except
    // If there's any error parsing the parameter, default to true
    Result := True;
  end;
end;


procedure TBaseInterceptor.LogRequestStart(const ARequest: TEndpointRequest; const AUserID: string);
begin
  TLogHelpers.LogMessage(Format('[%s] %s %s - User: %s - Started', [
    FRequestID, ARequest.MethodString, ARequest.BasePath, AUserID
    ]));
end;


procedure TBaseInterceptor.LogRequestComplete(const AStatusCode: Integer; const ADuration: Integer);
begin
  TLogHelpers.LogMessage(Format('[%s] Completed in %dms with status %d', [
    FRequestID, ADuration, AStatusCode
    ]));
end;

// IEMSResourceInterceptor implementation
procedure TBaseInterceptor.BeforeRequest(const AContext: TEndpointContext; const ARequest: TEndpointRequest; const AResponse: TEndpointResponse; var AHandled: Boolean);
var
  Token: string;
  IsAuthenticated: Boolean;
  UserID: string;
  IsValid: Boolean;
  ErrorMessage: string;
  ErrorResponse: TJSONObject;
begin
  // Initialize request tracking
  FRequestID := GenerateRequestID;
  FStartTime := Now;

  // Authentication
  Token := ExtractAuthToken(ARequest);
  DoCustomAuthentication(ARequest, IsAuthenticated, UserID);

  if not IsAuthenticated then
  begin
    if Token = '' then
    begin
      ErrorResponse := TJSONObject.Create;
      ErrorResponse.AddPair('error', 'Authentication required');
      ErrorResponse.AddPair('message', 'Missing X-Custom-Token header');
    end
    else
    begin
      ErrorResponse := TJSONObject.Create;
      ErrorResponse.AddPair('error', 'Authentication failed');
      ErrorResponse.AddPair('message', 'Invalid custom token');
    end;

    AResponse.StatusCode := 401;
    AResponse.Body.SetValue(ErrorResponse, True);
    AHandled := True;
    Exit;
  end;

  FCurrentUserID := UserID;

  // Custom validation
  DoCustomValidation(ARequest, IsValid, ErrorMessage);
  if not IsValid then
  begin
    ErrorResponse := TJSONObject.Create;
    ErrorResponse.AddPair('error', 'Validation failed');
    ErrorResponse.AddPair('message', ErrorMessage);

    AResponse.StatusCode := 400;
    AResponse.Body.SetValue(ErrorResponse, True);
    AHandled := True;
    Exit;
  end;

  // Log request start
  LogRequestStart(ARequest, FCurrentUserID);
end;


procedure TBaseInterceptor.AfterRequest(const AContext: TEndpointContext; const ARequest: TEndpointRequest; const AResponse: TEndpointResponse);
var
  Duration: Integer;
  StatusCode: Integer;
  IncludeMetadata: Boolean;
  ResponseObj, MetaObj: TJSONObject;
  ResponseValue: TJSONValue;
begin
  Duration := MillisecondsBetween(Now, FStartTime);

  StatusCode := AResponse.StatusCode;

  // Log request completion
  LogRequestComplete(StatusCode, Duration);

  // Check if metadata should be included
  IncludeMetadata := ShouldIncludeMetadata(ARequest);

  // Only add metadata for successful responses (2xx status codes)
  // Don't try to modify error responses as they may be in an invalid state
  if IncludeMetadata and (StatusCode >= 200) and (StatusCode < 300) then
  begin
    // Add metadata to response
    try
      // Create new response structure from scratch
      ResponseObj := TJSONObject.Create;

      if AResponse.Body.TryGetValue(ResponseValue) then
        ResponseObj.AddPair('data', ResponseValue)
      else
        ResponseObj.AddPair('data', TJSONNull.Create);

      // Add metadata
      MetaObj := TJSONObject.Create;
      MetaObj.AddPair('request_id', FRequestID);
      MetaObj.AddPair('timestamp', DateTimeToStr(Now));
      MetaObj.AddPair('duration_ms', TJSONNumber.Create(Duration));
      MetaObj.AddPair('user_id', FCurrentUserID);
      MetaObj.AddPair('method', ARequest.MethodString);
      MetaObj.AddPair('resource', ARequest.Resource);

      ResponseObj.AddPair('meta', MetaObj);

      // Set the new response body
      AResponse.Body.SetValue(ResponseObj, False);
    except
      on E: Exception do
      begin
        // If something goes wrong, just log it but don't break the response
        TLogHelpers.LogMessage(Format('[%s] Error adding metadata: %s', [FRequestID, E.Message]));
      end;
    end;
  end;
  // If metadata is not requested or response is an error, leave the response as-is

  // Allow descendants to add custom logging
  DoCustomLogging(ARequest, AResponse);
end;

// Default implementations - descendants can override these
procedure TBaseInterceptor.DoCustomAuthentication(const ARequest: TEndpointRequest; var AAuthenticated: Boolean; var AUserID: string);
var
  Token: string;
begin
  Token := ExtractAuthToken(ARequest);
  AAuthenticated := ValidateAuthToken(Token);
  if AAuthenticated then
    AUserID := GetUserIDFromToken(Token)
  else
    AUserID := '';
end;


procedure TBaseInterceptor.DoCustomValidation(const ARequest: TEndpointRequest; var AValid: Boolean; var AErrorMessage: string);
var
  ContentType: string;
begin
  // Default validation - just check content type for POST/PUT
  AValid := True;
  if (ARequest.Method = TEndpointRequest.TMethod.Post) or (ARequest.Method = TEndpointRequest.TMethod.Put) then
  begin
    ContentType := GetHeaderValue(ARequest, 'Content-Type');
    if ContentType = '' then
    begin
      AValid := False;
      AErrorMessage := 'Content-Type header is required';
    end
    else if not ContentType.StartsWith('application/json') then
    begin
      AValid := False;
      AErrorMessage := 'Content-Type must be application/json';
    end;
  end;
end;


procedure TBaseInterceptor.DoCustomLogging(const ARequest: TEndpointRequest; const AResponse: TEndpointResponse);
begin
  // Default implementation does nothing - descendants can override
end;

end.
